Virtual methods in interfaces

BillWagner opened the pull request:
Fixes #12418 
Fixes #12488 

Depends on dotnet/samples#1615

This tutorial explains how you can use default interface methods to declare virtual methods in interfaces. These enable you to create "mixins" that let you declare functionality by declaring which interfaces a type supports.

BillWagner committed:
partial first draft.

BillWagner committed:
interim checkin.

First draft almost done.

BillWagner committed:
checkpoint

BillWagner committed:
first draft is finished.

BillWagner committed:
Review and proofread.

gewarren commented on docs/csharp/tutorials/virtual-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+In this article, you'll learn how virtual default interface implementations enable new scenarios.
+
+## Design the application.
```
```suggestion
## Design the application
```

gewarren commented on docs/csharp/tutorials/virtual-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+## Limitations of extension methods
+
+One way you can implement behavior that appears as part of an interface is to define [extension methods](../programming-guide/classes-and-structs/extension-methods.md) that provide the default behavior. Interfaces declare a minimum set of members while providing a greater surface area for any class that implements that interface. The extension methods in <xref:System.Linq.Enumerable> provide the  ...
```
```suggestion
One way you can implement behavior that appears as part of an interface is to define [extension methods](../programming-guide/classes-and-structs/extension-methods.md) that provide the default behavior. Interfaces declare a minimum set of members while providing a greater surface area for any class that implements that interface. For example, the extension methods in <xref:System.Linq.Enumerable> provide the implementation for any sequence to be the source of a LINQ query.
```

gewarren commented on docs/csharp/tutorials/virtual-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+[!code-csharp[First overhead light](~/samples/csharp/tutorials/virtual-interface-methods/UnusedExampleCode.cs?name=SnippetOverheadLightV1)]
+
+In this tutorial, the code doesn't drive IoT devices, but emulate those activities by writing messages to the console. You can explore the code without automating your house.
```
```suggestion
In this tutorial, the code doesn't drive IoT devices, but emulates those activities by writing messages to the console. You can explore the code without automating your house.
```

dasMulli commented:
This caught my eye when skimming over it because it seems to suggest that interface methods by default aren't virtual. However, interface methods are always virtual regardless of having a default implementation or not (and DIMs are always virtual regardless if the `virtual` keyword is used or not).
So I'm just afraid this may ask learners to ask themselves "so how do I create non-virtual DIMs?".

gewarren commented on docs/csharp/tutorials/virtual-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+These changes compile cleanly, even though the `ExtraFancyLight` declares support for the `ILight` interface and both derived interfaces, `ITimerLight` and `IBlinkingLight`. There's only one "closest" implementation declared in the `ILight` interface. Any class that declared an override would become the one "closest" implementation. You saw examples in the preceding classes that overrode the memb ...
+
+Avoid overriding the same method in multiple derived interfaces. Doing so creates an ambiguous method call whenever a class implements both derived interfaces. For example, if both the `IBlinkingLight` and `ITimerLight` implemented an override of `PowerStatus`, the `OverheadLight` would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the ...
```
```suggestion
Avoid overriding the same method in multiple derived interfaces. Doing so creates an ambiguous method call whenever a class implements both derived interfaces. For example, if both the `IBlinkingLight` and `ITimerLight` implemented an override of `PowerStatus`, the `OverheadLight` would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the two derived interfaces. You can usually avoid this situation by keeping interface definitions small and focused on one feature. In this scenario, each capability of a light is its own interface; multiple interfaces are only inherited by classes.
```

gewarren reviewed (commented):
Was just looking through PRs randomly and have a couple suggestions.

BillWagner commented:
@dasMulli Thanks for starting this conversation. It sparks some interesting thoughts.

Interface methods are not virtual.  The sections of the spec on [Interface implementation inheritance](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/interfaces#interface-implementation-inheritance) and [Interface re-implementation](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/interfaces#interface-re-implementation) cover the behavior in detail.

That said, I like your suggestion to augment this with how to create non-virtual DIM implementations. I touch on the subject [in the tutorial on versioning](https://docs.microsoft.com/dotnet/csharp/tutorials/default-interface-members-versions#upgrade-with-default-interface-members), but I don't think that's enough.

Where and how can we make this subtlety more clear? How best can we make the distinction between implementing an interface and overriding virtual methods in a base class more clear?

BillWagner committed:
Apply suggestions from code review

Co-Authored-By: Genevieve Warren <gewarren@microsoft.com>

dasMulli commented:
it's complicated. I'm not sure if we're now talking about the interface instance method itself - which is always emitted as `.method public hidebysig newslot virtual instance` to IL - or the method on the implementing class that the interface is mapped to (which would need to be marked as `virtual` for overwriting in subclasses if my brain now doesn't fail me completely).
Granted, this is super unintuitive because I can override a DIM in a "sub-interface" without explicitly marking it as `virtual`, because interface instance methods are always emitted as virtual (but not as `abstract`)

> In addition, because default interface methods can be virtual, the method dispatch always chooses the implementation in the class.

This suggests to me that using the `virtual` keyword inside the interface definition would change the semantics.

dasMulli commented:
I might as well be missing the key statement here and can't see the wood through all the trees 🙃

BillWagner commented:
@dasMulli  I just ran some tests on the code. I've got some misleading statements in how I explain the concepts here.

I'll update before it's ready for a full review. Thanks for the discussion.

BillWagner committed:
edits on `virtual`

Its use in default interface methods is to add the method to the public interface of implementors.

BillWagner committed:
Change file name

BillWagner commented:
I updated the focus and the content. It's ready for review.  Thanks for the discussion @dasMulli

BillWagner committed:
update link folders.

dasMulli commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,122 @@
+
+Extension methods are resolved at compile time, using the declared type of the variable. Classes that implement the interface can provide a better implementation for any extension method. Variable declarations must match the implementing type to enable the compiler to choose that implementation. When the compile-time type matches the interface, method calls resolve to the extension method. Anothe ...
+
+Starting with C# 8.0, you can declare the default implementations as interface methods. Then, every class automatically uses the default implementation. Any class that can provide a better implementation can override the interface method definition with a better algorithm. In one sense, this technique sounds similar to how you could use [extension methods](../programming-guide/classes-and-structs ...
```
Nit: `WIthout`.

Super sorry for being persistent, but maybe you can help me clear up possible misunderstandings of the terminology on my side.

> WIthout the `virtual` keyword on the interface method, default interface methods are accessible only with a reference to the interface type.

From my understanding of reading this it would suggest that by adding the `virtual` keyword should make the commented out line in the following code compile (by interpreting the inverse):
```csharp
public interface ICanReturnMyName
{
    public virtual string MyName() => nameof(ICanReturnMyName);
}
public class NameReturner : ICanReturnMyName {}
class Program
{
    static void Main(string[] args)
    {
        NameReturner nr = new NameReturner();
        ICanReturnMyName iNr = nr;
        // next line doesn't compile
        //Console.WriteLine($"NameReturner:         {nr.MyName()}");
        Console.WriteLine($"ICanReturnMyName iNr: {iNr.MyName()}");
    }
}
```

BillWagner replied:
@dasMulli You are correct. Thanks for catching this. I honestly don't know what I was looking at when I tested this.  Fixing now.

BillWagner committed:
Fix some technical errors.

dasMulli commented on docs/toc.yml:
```diff
@@ -580,8 +580,10 @@
     - name: Advanced scenarios for string Interpolation
       href: csharp/tutorials/string-interpolation.md
-    - name: Safely update interfaces with default interface members
+    - name: Safely update interfaces with default interface methods
```
was that supposed to go in?

BillWagner replied:
Yes. The official feature name is "default interface methods", not "default interface members".  See [spec proposals](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/default-interface-methods.md)

dasMulli replied:
Ah yeah, I think the referenced file contains "members" as well.
"members" was also used by engineers on GitHub a few times when ppl asked for what they mean with "DIM" so that stuck in my head. (like half a year ago)

dasMulli commented:
I'd suggest to also remove the `virtual` keywords from the interface declarations in the sample code.

mairaw committed:
Merge branch 'master' into virtual-methods-in-interfaces

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+Start by creating the interface that defines the behavior for all lights:
+
+[!code-csharp[Declare base interface](~/samples/csharp/tutorials/mixins-with-interfaces/UnusedSampleCode.cs?name=SnippetILightInterfaceV1)]
```
This is missing. Perhaps:

```suggestion
[!code-csharp[Declare base interface](~/samples/csharp/tutorials/mixins-with-interfaces/UnusedExampleCode.cs?name=SnippetILightInterfaceV1)]
```

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+In this tutorial, the code doesn't drive IoT devices, but emulates those activities by writing messages to the console. You can explore the code without automating your house.
+
+Next, let's define the interfaces for a light that can automatically turn off after a timeout:
```
should this be singular?

```suggestion
Next, let's define the interface for a light that can automatically turn off after a timeout:
```

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+You could add a basic implementation to the overhead light, but a better solution is to modify this interface definition to provide a `virtual` default implementation:
+
+[!code-csharp[Timer interface](~/samples/csharp/tutorials/mixins-with-interfaces/ITimerLight.cs?name=SnippetTimerLightFinal)]
```
minor nit: this code has a sentence that says:
"Using default interface implementation for ITimerLight.TurnOnFor"

wondering if it would be better as "Using the default interface implementation for ITimerLight.TurnOnFor" or
"Using the default interface implementation for the ITimerLight.TurnOnFor method"

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+[!code-csharp[Timer interface](~/samples/csharp/tutorials/mixins-with-interfaces/ITimerLight.cs?name=SnippetTimerLightFinal)]
+
+By adding that change, the `OverheadLight` can implement the timer function by declaring support for the interface:
```
```suggestion
By adding that change, the `OverheadLight` class can implement the timer function by declaring support for the interface:
```

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+The advantages of default interface methods become clearer as you introduce more advanced capabilities. Using interfaces enables you to mix and match capabilities. It also enables each class author to choose between the default implementation and a custom implementation. Let's add an interface with a default implementation for a blinking light:
+
+[!code-csharp[Define the blinking light interface](~/samples/csharp/tutorials/mixins-with-interfaces/IBlinkingLight.cs?name=SnippetBlinkingLight)]
```
same nit as before for the code + uppercase Using on the first WriteLine call

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+A new light type, the `LEDLight` supports both the timer function and the blink function directly. This light style implements both the `ITimerLight` and `IBlinkingLight` interfaces, and overrides the `Blink` method:
+
+[!code-csharp[Override the blink function](~/samples/csharp/tutorials/mixins-with-interfaces/LEDLight.cs?name=SnippetOverheadLight)]
```
this code is blank on the page. should it be

```suggestion
[!code-csharp[Override the blink function](~/samples/csharp/tutorials/mixins-with-interfaces/LEDLight.cs?name=SnippetLEDLight)]
```

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+Avoid overriding the same method in multiple derived interfaces. Doing so creates an ambiguous method call whenever a class implements both derived interfaces. For example, if both the `IBlinkingLight` and `ITimerLight` implemented an override of `PowerStatus`, the `OverheadLight` would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the ...
+
+This sample shows one scenario where you can define discrete features that can be mixed into classes. You declare any set of supported functionality by declaring which interfaces a class supports. The use of virtual default interface methods enables classes to use or define a different implementation for any or all the interface methods. This language capability provides new ways to model the rea ...
```
adding hyphen to real world 

```suggestion
This sample shows one scenario where you can define discrete features that can be mixed into classes. You declare any set of supported functionality by declaring which interfaces a class supports. The use of virtual default interface methods enables classes to use or define a different implementation for any or all the interface methods. This language capability provides new ways to model the real-world systems you're building. Default interface methods provide a clearer way to express related classes that may mix and match different features using virtual implementations of those capabilities.
```

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+[!code-csharp[Override the blink and timer function](~/samples/csharp/tutorials/mixins-with-interfaces/ExtraFancyLight.cs?name=SnippetExtraFancyLight)]
+
+The `HalogenLight` you created earlier does not support blinking. You just don't add the `IBlinkingLight` to the list of its supported interfaces.
```
thought it could read better by being more direct

```suggestion
The `HalogenLight` you created earlier doesn't support blinking. So, don't add the `IBlinkingLight` to the list of its supported interfaces.
```

mairaw commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+These changes compile cleanly, even though the `ExtraFancyLight` declares support for the `ILight` interface and both derived interfaces, `ITimerLight` and `IBlinkingLight`. There's only one "closest" implementation declared in the `ILight` interface. Any class that declared an override would become the one "closest" implementation. You saw examples in the preceding classes that overrode the memb ...
+
+Avoid overriding the same method in multiple derived interfaces. Doing so creates an ambiguous method call whenever a class implements both derived interfaces. For example, if both the `IBlinkingLight` and `ITimerLight` implemented an override of `PowerStatus`, the `OverheadLight` would need to provide a more specific override. Otherwise, the compiler can't pick between the implementations in the ...
```
If the compiler can't pick, what happens? Not clear.

mairaw reviewed (approved):
Neat article @BillWagner. Left some comments for you to address before merging though.

pkulikov commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+---
+# Tutorial: Mix in functionality when creating classes using interfaces with default interface methods
+
+Beginning with C# 8.0 on .NET Core 3.0, you can define an implementation when you declare a member of an interface. This feature provides new capabilities where you can define default implementations for features declared in interfaces. Classes can pick when to override functionality, when to use the default functionality, and when not to declare support for discrete features.
```
@mairaw as far as I understand, this feature requires the updated runtime, which is not included in .NET Framework, for example, but only in .NET Core 3.0. I don't know if other .NET implementations would be updated to support the DIM feature. If my understanding is correct, @BillWagner should docs mention that reliance on the runtime? 

In general, it seems that customers would like to see the list of what C# 8.0 features rely on the updated runtime and which are not.

BillWagner replied:
I'm going to leave this language as written.

@mairaw Default interface members will not work, except on .NET Core 3.0. As @pkulikov mentioned, this feature relies on CLR changes that are part of .NET Core.

@pkulikov We're still discussing how best to explain the issues for C# 8.0 on other versions. There are dependencies on types, which is reasonably well understood and handled with .NET Standard. There are CLR dependencies (for Default interface members). There's also a new concern regarding nullable reference types. While the compiler implements the feature, it relies on metadata in libraries to provide better and more accurate warnings. The libraries for .NET framework, older .NET standard and older .NET Core versions do not have those annotations. They are "nullable oblivious", So it's complicated.

BillWagner committed:
Apply suggestions from code review

Co-Authored-By: Maira Wenzel <mairaw@microsoft.com>

BillWagner committed:
respond to feedback

dasMulli commented on docs/csharp/tutorials/mixins-with-default-interface-methods.md:
```diff
@@ -0,0 +1,121 @@
+
+[!code-csharp[pure Timer interface](~/samples/csharp/tutorials/mixins-with-interfaces/UnusedExampleCode.cs?name=SnippetPureTimerInterface)]
+
+You could add a basic implementation to the overhead light, but a better solution is to modify this interface definition to provide a `virtual` default implementation:
```
@BillWagner I think this `virtual` should also be removed..

dasMulli commented:
How do you feel about removing the `virtual` declarations from the interfaces in the sample code?

BillWagner commented:
@dasMulli Thanks for the reminder. New PR coming....
